File: java/ch17/src/test/java/com/thesoftwaregorilla/tdd/money/MoneyTest.java
package com.thesoftwaregorilla.tdd.money;

import org.junit.jupiter.api.*;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.CsvSource;


import java.util.HashMap;
import java.util.function.BiFunction;
import java.util.function.Function;

import static org.junit.jupiter.api.Assertions.*;

//<editor-fold desc="TO DO List">
//TODO: (DONE) Add tests to deal with division by zero exceptions in the Money class caused by no exchange rate set.
//TODO: (DONE) Clean up the arithmetic tests align augend and addend with from and to currency pairs.
//TODO: (DONE) Refactor these tests into a nested test structure for the Money class. Basic tests first, then currency arithmetic and conversions.
//</editor-fold>


@DisplayName("Money Class Tests")
public class MoneyTest {

    //<editor-fold desc="Test Setup">
    private static HashMap<String, Function<Integer, Money>> currencyFactories;
    private static final int STD_AMOUNT = 5;
    private static Bank bank;

    @BeforeAll
    public static void beforeAllSetUp() {
        getCurrencyFactories();
        bank = BankTest.getBankWithRates();
    }

    private static HashMap<String, Function<Integer, Money>> getCurrencyFactories() {
        if(currencyFactories == null) {
            currencyFactories = new HashMap<>();
            currencyFactories.put("USD", Money::dollar);
            currencyFactories.put("CHF", Money::franc);
            currencyFactories.put("ZAR", Money::rand);
        }
        return currencyFactories;
    }

    public static Function<Integer, Money> getCurrencyFactory(String currency) {
        return getCurrencyFactories().get(currency);
    }
    //</editor-fold>

    @Nested
    @DisplayName("construction")
    public class ConstructionTests {

        @DisplayName("constructor")
        @ParameterizedTest(name = "with currency = \"{0}\"")
        @CsvSource({
                "USD",
                "CHF",
                "ZAR"
        })
        public void testConstruction(String currency) {
            Money money = currencyFactories.get(currency).apply(STD_AMOUNT);
            assertNotNull(money);
            assertEquals(STD_AMOUNT, money.getAmount());
            assertEquals(currency, money.getCurrency());
        }

        @DisplayName("toString")
        @ParameterizedTest(name = "with currency = \"{0}\", amount = {1}")
        @CsvSource({
                "USD, 5",
                "CHF, 10",
                "ZAR, 20"
        })
        public void testToString(String currency, int amount) {
            assertEquals(amount + " " + currency, currencyFactories.get(currency).apply(amount).toString());
        }

        @DisplayName("equality")
        @ParameterizedTest(name = "with currency = \"{0}\", unequal currency = \"{1}\", unequal amount = {2}")
        @CsvSource({
                "USD, CHF, 6",
                "CHF, ZAR, 7",
                "ZAR, USD, 8"
        })
        public void testEquality(String currency, String unequalCurrency, int unequalAmount) {
            Money money = currencyFactories.get(currency).apply(STD_AMOUNT);
            assertEquals(money, currencyFactories.get(currency).apply(STD_AMOUNT));
            assertNotEquals(money, currencyFactories.get(currency).apply(unequalAmount));
            assertNotEquals(money, currencyFactories.get(unequalCurrency).apply(STD_AMOUNT));
        }

        @DisplayName("currency")
        @ParameterizedTest(name = "with \"{0}\"")
        @CsvSource ({
                "USD",
                "CHF",
                "ZAR"
        })
        public void testCurrency(String currency) {
            assertEquals(currency, currencyFactories.get(currency).apply(STD_AMOUNT).getCurrency());
        }
    }

    @Nested
    @DisplayName("arithmetic")
    public class ArithmeticTests {


        @Nested
        @DisplayName("multiply")
        public class MultiplicationTests {
            @DisplayName("test for aliasing")
            @ParameterizedTest(name = "with currency = \"{0}\", multiplier = {1},{2}")
            @CsvSource({
                    "USD, 2, 4",
                    "CHF, 3, 2",
                    "ZAR, 4, 3"
            })
            public void testMultiplication(String currency, int multiplier1, int multiplier2) {
                Money money = currencyFactories.get(currency).apply(STD_AMOUNT);
                assertEquals(new Money(STD_AMOUNT * multiplier1, currency), money.times(multiplier1));
                assertEquals(new Money(STD_AMOUNT * multiplier2, currency), money.times(multiplier2));
            }
        }

        @Nested
        @DisplayName("add")
        public class AdditionTests {
            @DisplayName("one currency")
            @ParameterizedTest(name = "with currency = \"{0}\"")
            @CsvSource({
                    "USD",
                    "CHF",
                    "ZAR"
            })
            public void testSimpleAddition(String currency) {
                Money money = currencyFactories.get(currency).apply(STD_AMOUNT);
                Expression sum = money.plus(money);
                Bank bank = new Bank();
                Money reduced = bank.reduce(sum, currency);
                assertEquals(new Money(STD_AMOUNT + STD_AMOUNT, currency), reduced);
            }
        }

        @Nested
        @DisplayName("reduce")
        public class ReductionTests {
            @Test
            @DisplayName("using Money class")
            public void testReduceMoney() {
                Money result = bank.reduce(Money.dollar(1), "USD");
                assertEquals(Money.dollar(1), result);
            }

            @Test
            @DisplayName("using Sum class")
            public void testReduceSum() {
                Expression sum = new Sum(Money.dollar(3), Money.dollar(4));
                Money result = bank.reduce(sum, "USD");
                assertEquals(Money.dollar(7), result);
            }

            @DisplayName("with mixed currency")
            @ParameterizedTest(name = "from \"{0}\", to \"{1}\". Rate = {2}, Amount = {3}")
            @CsvSource({
                    "CHF, USD, 2, 2",
                    "ZAR, USD, 17, 85",
                    "ZAR, CHF, 20, 60"
            })
            public void testReduceMoneyDifferentCurrency(String fromCurrency, String toCurrency, int rate, int amount) {
                Bank bank = new Bank();
                bank.addRate(fromCurrency, toCurrency, rate);
                Money result = bank.reduce(currencyFactories.get(fromCurrency).apply(amount), toCurrency);
                assertEquals(currencyFactories.get(toCurrency).apply(amount / rate), result);
            }
        }

        @Nested
        @DisplayName("mixed currency")
        public class ComplexArithmeticTests {

            @DisplayName("plus with Money")
            @ParameterizedTest(name = "from \"{0}\" {1}, to \"{2}\" {3}, expected = {4}")
            @CsvSource({
                    "CHF, 10, USD, 5, 10",
                    "ZAR, 20, CHF, 10, 11",
                    "ZAR, 34, USD, 5, 7",
                    "USD, 5, ZAR, 20, 0"
            })
            public void testMixedAddition(String from, int fromAmt, String to, int toAmt, int expected) {
                testReduceHarness(from, fromAmt, to, toAmt, expected,
                        (aug, add) -> add.plus(aug));
            }

            @DisplayName("plus with Sum")
            @ParameterizedTest(name = "from \"{0}\" {1}, to \"{2}\" {3}, expected = {4}")
            @CsvSource({
                    "CHF, 10, USD, 5, 15",
                    "ZAR, 20, CHF, 10, 21",
                    "ZAR, 34, USD, 5, 12",
                    "USD, 5, ZAR, 20, 0"
            })
            public void testSumPlusMoney(String from, int fromAmt, String to, int toAmt, int expected) {
                testReduceHarness(from, fromAmt, to, toAmt, expected,
                        (aug, add) -> new Sum(aug, add).plus(add));
            }


            @DisplayName("times")
            @ParameterizedTest(name = "from \"{0}\" {1}, to = \"{2}\" {3}, expected = {4}")
            @CsvSource({
                    "CHF, 10, USD, 5, 20",
                    "ZAR, 20, CHF, 10, 22",
                    "ZAR, 85, USD, 5, 20",
                    "USD, 5, ZAR, 20, 0"
            })
            public void testSumTimes(String from, int fromAmt, String to, int toAmt, int expected) {
                testReduceHarness(from, fromAmt, to, toAmt, expected,
                        (aug, add) -> new Sum(aug, add).times(2));
            }

            private void testReduceHarness(String from, int fromAmt, String to, int toAmt, int expected, BiFunction<Expression, Expression, Expression> operation) {
                Expression money1 = currencyFactories.get(from).apply(fromAmt);
                Expression money2 = currencyFactories.get(to).apply(toAmt);
                Expression sum = operation.apply(money1, money2);
                if (bank.rate(from, to) == 0) {
                    ArithmeticException exception = assertThrows(ArithmeticException.class, () -> {
                        Money result = bank.reduce(sum, to);
                    });
                    assertEquals("Exchange rate not available", exception.getMessage());
                    return;
                }
                Money result = bank.reduce(sum, to);
                assertEquals(currencyFactories.get(to).apply(expected), result);
            }
        }
    }

    @Nested
    @DisplayName("incomplete tests")
    public class IncompleteTests {
        // This test is discussed at the end of Chapter 16, but there is no clean implementation for how to fix make it pass in the book.
        // The test is commented out because it will fail. We'll come back to it later.
        @Test
        @DisplayName("same currency should return Money")
        @Disabled("Disabled until we implement the fix for duplicate plus implementation")
        public void testPlusSameCurrencyReturnsMoney() {
            Expression sum = Money.dollar(1).plus(Money.dollar(1));
            assertInstanceOf(Money.class, sum);
        }

    }

}

File: cs/ch17-01-tests/Money.Tests/MoneyTest.cs
using Microsoft.VisualStudio.TestPlatform.ObjectModel;
using NUnit.Framework;
using NUnit.Framework.Internal;
using TheSoftwareGorilla.TDD.Money;

namespace TheSoftwareGorilla.TDD.Money.Tests;

#region TO DO's
#endregion
[TestFixture]
[Description("MoneyTest class")]
public class MoneyTest
{
    
    private static readonly Dictionary<String, Func<int, Money>> _currencyFactories = new Dictionary<string, Func<int, Money>>
    {
        { "USD", Money.Dollar },
        { "CHF", Money.Franc },
        { "ZAR", Money.Rand }
    };
    private const int STD_AMT = 5;

    [OneTimeSetUp]
    public static void OneTimeSetUp()
    {
        // This method is called only once before any of the tests are run.
    }

    public static Func<int, Money> GetCurrencyFactory(string currency)
    {
        return _currencyFactories[currency];
    }

    private Money? _fiveDollar;
    
    [SetUp]
    public void Setup()
    {
        _fiveDollar = Money.Dollar(5);
    }

    [TearDown]
    public void TearDown()
    {
        _fiveDollar = null;
    }

    [TestCase("USD", 5, 5, TestName = "construct USD 5")]
    [TestCase("CHF", 5, 5, TestName = "construct CHF 5")]
    [TestCase("ZAR", 5, 5, TestName = "construct ZAR 5")]
    [Category("construction")]
    public void TestConstruction(string currency, int amount, int expected)
    {
        var money = _currencyFactories[currency].Invoke(amount);
        Assert.IsNotNull(money);
        Assert.That(money.Amount, Is.EqualTo(expected));
        Assert.That(money.Currency, Is.EqualTo(currency));
    }

    [TestCase("USD", 5, TestName = "ToString USD 5")]
    [TestCase("CHF", 5, TestName = "ToString CHF 10")]
    [TestCase("ZAR", 5, TestName = "ToString ZAR 20")]
    [Category("construction")]
    public void TestToString(string currency, int amount)
    {
        var money = _currencyFactories[currency].Invoke(amount);
        Assert.That(money.ToString(), Is.EqualTo(amount + " " + currency));
    }

    [TestCase("USD", "CHF",  6, TestName = "equality USD 6")]
    [TestCase("CHF", "ZAR", 7, TestName = "equality CHF 7")]
    [TestCase("ZAR", "USD", 8, TestName = "equality ZAR 8")]
    [Category("construction")]
    public void TestEquality(string currency1, string currency2, int notEqualAmount)
    {
        var money1 = _currencyFactories[currency1].Invoke(STD_AMT);
        var money2 = _currencyFactories[currency2].Invoke(STD_AMT);
        Assert.That(money1, Is.EqualTo(_currencyFactories[currency1].Invoke(STD_AMT)));
        Assert.That(money1, Is.Not.EqualTo(_currencyFactories[currency1].Invoke(notEqualAmount)));
        Assert.That(money2, Is.Not.EqualTo(money1));
    }

    [TestCase("USD", TestName = "currency USD")]
    [TestCase("CHF", TestName = "currency CHF")]
    [TestCase("ZAR", TestName = "currency ZAR")]
    [Category("construction")]
    public void TestCurrency(string currency)
    {
        var money = _currencyFactories[currency].Invoke(STD_AMT);
        Assert.That(money.Currency, Is.EqualTo(currency));
    }

    [TestCase("USD", 2, 4, TestName = "multiply test for aliasing USD, multiplier  2, 4")]
    [TestCase("CHF", 3, 2, TestName = "multiply test for aliasing CHF, multiplier 3, 2")]
    [TestCase("ZAR", 4, 3, TestName = "multiply test for aliasing ZAR, multiplier 4, 3")]
    [Category("multiplication")]
    public void TestMultiplication(string currency, int multiplier1, int multiplier2)
    {
        var money = _currencyFactories[currency].Invoke(STD_AMT);
        Assert.That(_currencyFactories[currency].Invoke(STD_AMT * multiplier1), Is.EqualTo(money.Times(multiplier1)));
        Assert.That(_currencyFactories[currency].Invoke(STD_AMT * multiplier2), Is.EqualTo(money.Times(multiplier2)));
    }

    [TestCase("USD", TestName = "simple add currency USD")]
    [TestCase("CHF", TestName = "simple add currency CHF")]
    [TestCase("ZAR", TestName = "simple add currency ZAR")]
    [Category("addition")]
    public void TestSimpleAddition(string currency)
    {
        var money = _currencyFactories[currency].Invoke(STD_AMT);
        Expression sum = money.Plus(money);
        Assert.That(sum, Is.InstanceOf<Sum>());
        Money reduced = BankTest.GetBankWithRates().Reduce(sum, currency);
        Assert.That(reduced, Is.EqualTo(_currencyFactories[currency].Invoke(STD_AMT + STD_AMT)));
    }


    [TestCase (TestName = "reduce using Sum class")]
    [Category("reduction")]
    public void TestReduceSum()
    {
        Expression sum = new Sum(Money.Dollar(3), Money.Dollar(4));
        Money result = BankTest.GetBankWithRates().Reduce(sum, "USD");
        Assert.That(Money.Dollar(7), Is.EqualTo(result));
    }

    [TestCase (TestName = "reduce using Money class")]
    [Category("reduction")]
    public void TestReduceMoney()
    {
        Money result = BankTest.GetBankWithRates().Reduce(Money.Dollar(1), "USD");
        Assert.That(Money.Dollar(1), Is.EqualTo(result));
    }

    [TestCase ("CHF", "USD", 2, 2, TestName = "reduce money mixed currency from CHF to USD, rate 2, amount 2")]
    [TestCase ("ZAR", "USD", 17, 85, TestName = "reduce money mixed currency from ZAR to USD, rate 17, amount 85")]
    [TestCase ("ZAR", "CHF", 20, 60, TestName = "reduce money mixed currency from ZAR to CHF, rate 20, amount 60")]
    [Category("reduction")]
    public void TestReduceMoneyDifferentCurrency(string fromCurrency, string toCurrency, int rate, int amount)
    {
        Bank bank = new Bank();
        bank.AddRate(fromCurrency, toCurrency, rate);
        Money result = bank.Reduce(_currencyFactories[fromCurrency].Invoke(amount), toCurrency);
        Assert.That(_currencyFactories[toCurrency].Invoke(amount / rate), Is.EqualTo(result));
    }


    [TestCase ("CHF", 10, "USD", 5, 10, TestName = "plus with Money mixed currency CHF 10 with USD 5, expected 10")]
    [TestCase ("ZAR", 20, "CHF", 10, 11, TestName = "plus with Money mixed currency ZAR 20 with CHF 10, expected 11")]
    [TestCase ("ZAR", 34, "USD", 5, 7, TestName = "plus with Money mixed currency ZAR 34 with USD 5, expected 7")]
    [TestCase ("USD", 5, "ZAR", 20, 0, TestName = "plus with Money missing exchange rate USD 5 with ZAR 20, expected InvalidOperationException")]
    [Category("complex arithmetic")]
    public void TestMixedAddition(string fromCurrency, int fromAmount, string toCurrency, int toAmount, int expected)
    {
        testReduceHarness(fromCurrency, fromAmount, toCurrency, toAmount, expected, (from, to) => from.Plus(to));
    }


    [TestCase ("CHF", 10, "USD", 5, 15, TestName = "plus with Sum mixed currency CHF 10 with USD 5, expected 15")]
    [TestCase ("ZAR", 20, "CHF", 10, 21, TestName = "plus with Sum mixed currency ZAR 20 with CHF 10, expected 21")]
    [TestCase ("ZAR", 34, "USD", 5, 12, TestName = "plus with Sum mixed currency ZAR 34 with USD 5, expected 12")]
    [TestCase ("USD", 5, "ZAR", 20, 0, TestName = "plus with Sum missing exchange rate USD 5 with ZAR 20, expected InvalidOperationException")]
    [Category("complex arithmetic")]
    public void TestSumPlusMoney(string fromCurrency, int fromAmount, string toCurrency, int toAmount, int expected)
    {
        testReduceHarness(fromCurrency, fromAmount, toCurrency, toAmount, expected, (from, to) => new Sum(from, to).Plus(to));
    }

    [TestCase ("CHF", 10, "USD", 5, 20, TestName = "times with mixed currency CHF 10 with USD 5, expected 20")]
    [TestCase ("ZAR", 20, "CHF", 10, 22, TestName = "times with mixed currency ZAR 20 with CHF 10, expected 22")]
    [TestCase ("ZAR", 85, "USD", 5, 20, TestName = "times with mixed currency ZAR 85 with USD 5, expected 20")]
    [TestCase ("USD", 5, "ZAR", 20, 0, TestName = "times with missing exchange rate USD 5 with ZAR 20, expected InvalidOperationException")]
    [Category("complex arithmetic")]
    public void TestSumTimes(string fromCurrency, int fromAmount, string toCurrency, int toAmount, int expected)
    {
        testReduceHarness(fromCurrency, fromAmount, toCurrency, toAmount, expected, (from, to) => new Sum(from, to).Times(2));
    }
    
    private static void testReduceHarness(string from, int fromAmt, string to, int toAmt, int expected, Func<Expression, Expression, Expression> operation)
    {
        Expression fromMoney = _currencyFactories[from].Invoke(fromAmt);
        Expression toMoney = _currencyFactories[to].Invoke(toAmt);
        Expression sum = operation.Invoke(fromMoney, toMoney);
        Bank bank = BankTest.GetBankWithRates();
        if (bank.Rate(from, to) == 0)
        {
            Assert.Throws<InvalidOperationException>(() => bank.Reduce(sum, to));
            return;
        }
        Money result = BankTest.GetBankWithRates().Reduce(sum, to);
        Assert.That(result, Is.EqualTo(_currencyFactories[to].Invoke(expected)));
    }

    // This test is discussed at the end of Chapter 16, but there is no clean implementation for how to fix make it pass in the book.
    // The test is commented out because it will fail. We'll come back to it later.
    [Test]
    [Ignore("Disabled until we implement the fix for duplicate plus implementation")]
    public void TestPlusSameCurrencyReturnsMoney()
    {
        Expression sum = Money.Dollar(1).Plus(Money.Dollar(1));
        Assert.That(sum, Is.InstanceOf<Money>());
    }

}